### PR TITLE
Custom extension classes on Kafka Connect role

### PR DESCRIPTION
**Pull request by jeqo**
# Description

To deploy Confluent Replicator in a Connector cluster, monitoring extensions have to be added. Currently, this configuration is only modified when RBAC is enabled. 

This PR proposes a new variables to be concatenated when defined.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

**Test Configuration**: molecule/rbac-scram-custom-rhel

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.test/molecule/rbac-scram-custom-rhel/molecule.yml [hunk 1]
```diff
@@ -146,8 +146,8 @@ platforms:
     privileged: true
     ## Commenting below to avoid port collision on jenkins server
     ## Uncomment to view c3 in browser at localhost:9021
-    # published_ports:
-    #   - "9021:9021"
+    published_ports:
```
Comment (domenicbove) [inline roles/confluent.test/molecule/rbac-scram-custom-rhel/molecule.yml:149]: keep these commented out

Comment (jeqo) [reply roles/confluent.test/molecule/rbac-scram-custom-rhel/molecule.yml:149]: 😓 always forgot this

File: roles/confluent.variables/defaults/main.yml [hunk 2]
```diff
@@ -529,6 +529,10 @@ kafka_connect_key_path: "/var/ssl/private/kafka_connect.key"
 kafka_connect_export_certs: "{{kafka_connect_ssl_mutual_auth_enabled}}"
 kafka_connect_keytab_path: /etc/security/keytabs/kafka_connect.keytab
 kafka_connect_kafka_listener_name: internal
+
+### Additional extension classes to be added to
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:533]: This comment should say what format this variable needs to be in, It appears it should be a comma separated string
Do you like comma separated string over list?

also run python doc.py to get the variable into variables.md

File: roles/confluent.variables/vars/main.yml [hunk 3]
```diff
@@ -504,6 +504,7 @@ kafka_connect_properties:
       producer.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
       consumer.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
       confluent.license.topic: _confluent-command
+      rest.extension.classes: "{% if kafka_connect_extension_classes_custom | length >0 %}{{ kafka_connect_extension_classes_custom }}{% endif %}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:507]: I think in the case of non rbac, and an unchanged kafka_connect_extension_classes_custom, then the dictionary will then have `rest.extension.classes: ""` which we don't want. You can do this

```suggestion
  rest_classes:
      enabled: "{{ kafka_connect_extension_classes_custom||length > 0 }}"
      rest.extension.classes: "{{ kafka_connect_extension_classes_custom }}"
```

**Comment by domenicbove** [comment]
also there are linting errors, you can run
```
yamllint -c .yamllint .[0m
```

File: roles/confluent.variables/defaults/main.yml (hunk 2 above)
Comment (jeqo) [reply roles/confluent.variables/defaults/main.yml]: Good point. I'd say in this case comma-separated over list is simpler.

File: roles/confluent.variables/vars/main.yml (hunk 3 above)
Comment (jeqo) [reply roles/confluent.variables/vars/main.yml]: seems to require a properties object after enabled, following the other lists.

File: roles/confluent.variables/vars/main.yml [hunk 4]
```diff
@@ -580,7 +584,7 @@ kafka_connect_properties:
   rbac:
     enabled: "{{rbac_enabled}}"
     properties:
-      rest.extension.classes: "io.confluent.connect.security.ConnectSecurityExtension{% if kafka_connect_secret_registry_enabled|bool %},io.confluent.connect.secretregistry.ConnectSecretRegistryExtension{% endif %}"
+      rest.extension.classes: "{% if kafka_connect_extension_classes_custom | length >0 %}{{ kafka_connect_extension_classes_custom }},{% endif %}io.confluent.connect.security.ConnectSecurityExtension{% if kafka_connect_secret_registry_enabled|bool %},io.confluent.connect.secretregistry.ConnectSecretRegistryExtension{% endif %}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:587]: this if statement isnt necessary right? if the string has length 0 then just concatenate it

File: roles/confluent.variables/vars/main.yml [hunk 5]
```diff
@@ -504,6 +504,10 @@ kafka_connect_properties:
       producer.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
       consumer.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
       confluent.license.topic: _confluent-command
+  rest_classes:
+    enabled: "{{ kafka_connect_extension_classes_custom||length > 0 }}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:508]:
```suggestion
    enabled: "{{ kafka_connect_extension_classes_custom|length > 0 }}"
```

File: roles/confluent.variables/vars/main.yml (hunk 4 above)
Comment (jeqo) [reply roles/confluent.variables/vars/main.yml:587]: its only purpose is to keep the comma in place after the custom classes.

Comment (domenicbove) [reply roles/confluent.variables/vars/main.yml:587]: Oh you're so right!

Comment (domenicbove) [reply roles/confluent.variables/vars/main.yml:587]: I'm not really enjoying all the if statements, this is a solution i came up with on a different PR that uses list objects:

```
# in roles/confluent.variables/defaults/main.yml
### Additional set of comma-separated extension classes
kafka_connect_extension_classes_custom: ""


# in roles/confluent.variables/vars/main.yml
kafka_connect_rest_extension_classes:
  - "{% if rbac_enabled|bool %}io.confluent.connect.security.ConnectSecurityExtension{% endif %}"
  - "{% if kafka_connect_secret_registry_enabled|bool %},io.confluent.connect.secretregistry.ConnectSecretRegistryExtension{% endif %}"
 
kafka_connect_final_rest_extension_classes: "{{kafka_connect_rest_extension_classes| difference(['']) + kafka_connect_custom_rest_extension_classes.split(',')}}"

  rest_classes:
    enabled: "{{ kafka_connect_final_rest_extension_classes|length > 0 }}"
    properties:
      rest.extension.classes: "{{ kafka_connect_final_rest_extension_classes.join(',') }}"
```

WDYT?

**Comment by domenicbove** [comment]
@jeqo What are your thoughts on the last comment. I'd like to get this PR thru soon because it is so close

**Comment by jeqo** [comment]
@domenicbove I like the list approach better now that it covers the other properties as well. Will give a try later today.

File: roles/confluent.variables/vars/main.yml [hunk 6]
```diff
@@ -475,6 +475,12 @@ kafka_connect:
 
 kafka_connect_http_protocol: "{{ 'https' if kafka_connect_ssl_enabled|bool else 'http' }}"
 
+kafka_connect_rest_extension_classes:
+  - "{% if rbac_enabled|bool %}io.confluent.connect.security.ConnectSecurityExtension{% endif %}"
+  - "{% if kafka_connect_secret_registry_enabled|bool %},io.confluent.connect.secretregistry.ConnectSecretRegistryExtension{% endif %}"
+ 
+kafka_connect_final_rest_extension_classes: "{{kafka_connect_rest_extension_classes| difference(['']) + kafka_connect_custom_rest_extension_classes.split(',')}}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:482]:
```suggestion
kafka_connect_final_rest_extension_classes: "{{ (kafka_connect_rest_extension_classes | difference(['']) + kafka_connect_custom_rest_extension_classes.split(',')) | unique}}"
```

We should probably add the unique filter just to be safe

File: roles/confluent.variables/vars/main.yml [hunk 7]
```diff
@@ -505,9 +511,13 @@ kafka_connect_properties:
       consumer.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
       confluent.license.topic: _confluent-command
   rest_classes:
+<<<<<<< HEAD
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:514]: typo

File: roles/confluent.variables/defaults/main.yml [hunk 8]
```diff
@@ -529,6 +529,10 @@ kafka_connect_key_path: "/var/ssl/private/kafka_connect.key"
 kafka_connect_export_certs: "{{kafka_connect_ssl_mutual_auth_enabled}}"
 kafka_connect_keytab_path: /etc/security/keytabs/kafka_connect.keytab
 kafka_connect_kafka_listener_name: internal
+
+### Additional set of extension classes
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:533]:
```suggestion
### Additional set of Connect extension classes.
```

File: VARIABLES.md [hunk 9]
```diff
@@ -964,6 +964,14 @@ Default:  "{{ ssl_mutual_auth_enabled }}"
 
 ***
 
+### kafka_connect_custom_rest_extension_classes
+
+Additional set of extension classes
```
Comment (domenicbove) [inline VARIABLES.md:969]:
```suggestion
Additional set of Connect extension classes.
```

File: VARIABLES.md [hunk 10]
```diff
@@ -966,7 +966,7 @@ Default:  "{{ ssl_mutual_auth_enabled }}"
 
 ### kafka_connect_custom_rest_extension_classes
 
-Additional set of extension classes
+Additional set of Connect extension classes.
```
Comment (domenicbove) [inline VARIABLES.md:969]: shoot, actually make this change in the defaults/main.yml and then rerun the docs script

**Review by domenicbove — APPROVED** [review]
Looks great

**Review by JumaX — APPROVED** [review]
LGTM.

**Pull request merged**